kyo-prelude/shared/src/main/scala/kyo2/Result.scala

```diff
@@ -4,134 +4,159 @@ import Result.*
-                case self: Failure[A] =>
-                    ifFailure(self.exception)
+                case self: Failure[E] =>
+                    ifFailure(self.error)
+                case self: Panic =>
+                    ifPanic(self.exception)
                 case _ =>
                     try
                         self match
-                            case self: SuccessFailure[A] @unchecked =>
-                                ifSuccess(self.unnest.asInstanceOf[A])
+                            case self: SuccessFailure[T] @unchecked =>
+                                ifSuccess(self.unnest.asInstanceOf[T])
                             case self =>
-                                ifSuccess(self.asInstanceOf[A])
+                                ifSuccess(self.asInstanceOf[T])
                     catch
-                        case ex if NonFatal(ex) => ifFailure(ex)
+                        case ex if NonFatal(ex) => ifPanic(ex)
 
-        inline def get: A =
-            fold(ex => throw ex)(identity)
+        def get(using E =:= Nothing): T =
+            fold(ex => throw new NoSuchElementException(s"Cause: $ex"))(throw _)(identity)
 
-        inline def getOrElse[B >: A](inline default: => B): B =
-            fold(_ => default)(identity)
+        inline def getOrElse[U >: T](inline default: => U): U =
+            fold(_ => default)(_ => default)(identity)
 
-        def orElse[B >: A](alternative: => Result[B]): Result[B] =
-            fold(_ => alternative)(v => Result.success(v))
+        def orElse[E2, U >: T](alternative: => Result[E2, U]): Result[E | E2, U] =
+            fold(_ => alternative)(_ => alternative)(v => Result.success(v))
 
-        inline def flatMap[B](inline f: A => Result[B]): Result[B] =
+        inline def flatMap[E2, U](inline f: T => Result[E2, U]): Result[E | E2, U] =
             (self: @unchecked) match
-                case _: Failure[?] =>
-                    self.asInstanceOf[Result[B]]
-                case _ =>
-                    try f(self.get)
+                case self: Cause[E] => self
+                case self =>
+                    try f(self.asInstanceOf[Success[T]].get)
                     catch
                         case ex if NonFatal(ex) =>
-                            Failure(ex)
+                            Panic(ex)
 
-        inline def flatten[B](using ev: A <:< Result[B]): Result[B] =
+        def flatten[E2, U](using ev: T <:< Result[E2, U]): Result[E | E2, U] =
             flatMap(ev)
 
-        inline def map[B](inline f: A => B): Result[B] =
+        inline def map[U](inline f: T => U): Result[E, U] =
             flatMap(v => Result.success(f(v)))
 
-        inline def filter(inline p: A => Boolean): Result[A] =
+        inline def withFilter(inline p: T => Boolean): Result[E | NoSuchElementException, T] =
+            filter(p)
+
+        inline def filter(inline p: T => Boolean): Result[E | NoSuchElementException, T] =
```

Comment:
I don't think the `E` is correct here - the Panic will not be visible in the result type.

Reply:
I meant to return `Failure`, not `Panic`. I've fixed it